working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+  X front() => this[0];
+}
+
+void main() {
+  ListSize<String> xs = <String>['Hello']; // OK, upcast.
+  print(xs); // OK, `toString()` available on Object?.
+  print("Size: ${xs.size}. Front: ${xs.front()}"); // OK.
+  xs[0]; // Error, `operator []` not a member of `ListSize`.
+
+  List<ListSize<String>> ys = [xs]; // OK.
+  List<List<String>> ys2 = ys; // Error, downcast.
+  ListSize<ListSize<Object>> ys3 = ys; // OK.
+  ys[0].front(); // OK.
+  ys3.front().front(); // OK.
+  ys as List<List<String>>; // `ys` is promoted, succeeds at run time.
+}
+```
+
+
+## Syntax
+
+The rule for `<extensionDeclaration>` in the grammar is replaced by the
+following:
+
+```ebnf
+<extensionDeclaration> ::=
+  'protected'? 'extension' ('type'? <typeIdentifier> <typeParameters>?)?
+      <extensionShowHidePart> 'on' <type> <interfaces>? '{'
+    (<metadata> <extensionMemberDefinition>)*
+  '}'
+
+<extensionShowHidePart> ::=
+  <extensionShowClause>? <extensionHideClause>?
+
+<extensionShowClause> ::= 'show' <extensionShowHideList>
+
+<extensionHideClause> ::= 'hide' <extensionShowHideList>
+
+<extensionShowHideList> ::=
+  <extensionShowHideElement> (',' <extensionShowHideElement>)*
+
+<extensionShowHideElement> ::=
+  <type> | <identifier> | 'operator' <operator>
+```
+
+*In the rule `<extensionShowHideElement>`, note that `<type>` derives
+`<typeIdentifier>`, which makes `<identifier>` nearly redundant. However,
+`<identifier>` is still needed because it includes some strings that cannot
+be the name of a type but can be the basename of a member, e.g., the
+built-in identifiers.*
+
+
+## Static Analysis
+
+Assume that _E_ is an extension declaration of the following form:
+
+```dart
+extension Ext<X1 extends B1, .. Xk extends Bk> on T {
+  ... // Members
+}
+```
+
+*Note that `extension type` declarations and other enhanced forms are
+discussed in later sections. The properties specified here are valid for
+those forms as well, except for the differences which are specified
+explicitly.*
+
+It is then allowed to use `Ext<S1, .. Sk>` as a type: It can occur as the
+declared type of a variable or parameter, as the return type of a function
+or getter, as a type argument in a type, or as the on-type of an extension.
+
+*In particular, it is allowed to create a new instance where one or more
+extension types occur as type arguments.*
+
+When `k` is zero, `Ext<S1, .. Sk>` simply stands for `Ext`, a non-generic
+extension. When `k` is greater than zero, a raw occurrence `Ext` is treated
+like a raw type: Instantiation to bound is used to obtain the omitted type
+arguments.
+
+We say that the static type of said variable, parameter, etc. _is the
+extension type_ `Ext<S1, .. Sk>`, and that its static type _is an extension
+type_.
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is treated as
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>>(v).m(args)</code>
```

Comment:
Small nit, but maybe worth nothing that `Ext` may not actually be in scope (or even denotable in the current scope, since it might be a private name from another library).

Reply:
Right.

I think this situation is similar to, for instance, `var x = foo();` where `foo` is a function imported from a different library `L` whose return type is a private type `_C` declared in `L`. In that situation we'd be able to access the members in the interface of `_C`, even though we cannot denote that type syntactically in this library. Similarly, we could have `var xs = [foo()];` where we'd infer and reify `_C` as the type argument of the list literal, as well as giving `xs` a declared type containing `_C`.

Another case is the implicitly induced noSuchMethod forwarders ['forced by privacy'](https://github.com/dart-lang/language/blob/53f3e4606f7efe25679c37698e167734df7f934c/specification/dartLangSpec.tex#L2825). Such a method cannot be written as Dart code, because it is required to override a method with a private name, and that method is declared in a different library.

So we already have several cases where we rely on the ability for a source code transformation to introduce references to private entities in different libraries, even though that cannot be expressed in Dart proper.